java-checks/src/main/java/org/sonar/java/checks/security/UserEnumerationCheck.java

```diff
@@ -0,0 +1,123 @@
+
+  private static final String MESSAGE = "Make sure allowing user enumeration is safe here.";
+  private static final String ABSTRACT_USER_DETAILS_AUTHENTICATION_PROVIDER = "org.springframework.security.authentication.dao.AbstractUserDetailsAuthenticationProvider";
+  private static final String USER_DETAILS_SERVICE = "org.springframework.security.core.userdetails.UserDetailsService";
+  private static final String USERNAME_NOT_FOUND_EXCEPTION = "org.springframework.security.core.userdetails.UsernameNotFoundException";
+  private static final String HIDE_USER_NOT_FOUND_EXCEPTIONS = "setHideUserNotFoundExceptions";
+  private static final String LOAD_USER_BY_USERNAME = "loadUserByUsername";
+  private static final String BOOLEAN = "boolean";
+  private static final String STRING = "java.lang.String";
+  private static final String THROWABLE = "java.lang.Throwable";
+
+  private final MethodMatchers setHideUserMatcher = MethodMatchers.create()
+    .ofSubTypes(ABSTRACT_USER_DETAILS_AUTHENTICATION_PROVIDER)
+    .names(HIDE_USER_NOT_FOUND_EXCEPTIONS)
+    .addParametersMatcher(BOOLEAN)
+    .build();
+
+  private final MethodMatchers loadUserMatcher = MethodMatchers.create()
+    .ofSubTypes(USER_DETAILS_SERVICE)
+    .names(LOAD_USER_BY_USERNAME)
+    .addParametersMatcher(STRING)
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.METHOD_INVOCATION, Tree.Kind.THROW_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.THROW_STATEMENT)) {
+      checkThrowUsernameNotFoundException(((ThrowStatementTree) tree));
+      return;
+    }
+
+    MethodInvocationTree methodInvocationTree = (MethodInvocationTree) tree;
+    Arguments arguments = methodInvocationTree.arguments();
+    if (arguments.isEmpty()){
+      return;
+    }
+    ExpressionTree expression = arguments.get(0);
+
+    checkHiddenUserNotFoundException(methodInvocationTree, expression);
+
+    checkLoadUserArgUsedInExceptions(methodInvocationTree, expression);
+  }
+
+  private void checkLoadUserArgUsedInExceptions(MethodInvocationTree methodInvocationTree, ExpressionTree expression) {
+    if (loadUserMatcher.matches(methodInvocationTree) && expression.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifierTree = (IdentifierTree) expression;
+      identifierTree.symbol().usages()
+        .stream().filter(UserEnumerationCheck::checkParentIsThrowable)
+        .forEach(value -> reportIssue(value, MESSAGE));
+    }
+  }
+
+  private void checkHiddenUserNotFoundException(MethodInvocationTree methodInvocationTree, ExpressionTree expression) {
+    if (setHideUserMatcher.matches(methodInvocationTree) && isFalseLiteral(expression)) {
+      reportIssue(methodInvocationTree, MESSAGE);
+    }
+  }
+
+  private void checkThrowUsernameNotFoundException(ThrowStatementTree tree) {
+    if (tree.expression().symbolType().is(USERNAME_NOT_FOUND_EXCEPTION)) {
```

Comment:
According to the rule description, it should be authorized to throw such an exception if we are within the body of the `loadUserByUsername` method. I don't see check over this case, and it would therefore report FPs.